plugins/inputs/windows_event_log/wineventlog/utils.go

```diff
@@ -50,39 +54,54 @@ func RenderEventXML(eventHandle EvtHandle, renderBuf []byte) ([]byte, error) {
 	return utf16ToUTF8Bytes(renderBuf, bufferUsed)
 }
 
-func CreateBookmark(channel string, recordID uint64) (h EvtHandle, err error) {
+func CreateBookmark(w WindowsEventAPI, channel string, recordID uint64) (h EvtHandle, err error) {
 	xml := fmt.Sprintf(bookmarkTemplate, channel, recordID)
 	p, err := syscall.UTF16PtrFromString(xml)
 	if err != nil {
 		return 0, err
 	}
-	h, err = EvtCreateBookmark(p)
+	h, err = w.EvtCreateBookmark(p)
 	if err != nil {
 		return 0, fmt.Errorf("error when creating a bookmark. Details: %v", err)
 	}
 	return h, nil
 }
 
 func CreateQuery(path string, levels []string) (*uint16, error) {
-	var filterLevels string
-	for _, level := range levels {
-		if filterLevels == "" {
-			filterLevels = fmt.Sprintf(eventLogLevelFilter, level)
-		} else {
-			filterLevels = filterLevels + " or " + fmt.Sprintf(eventLogLevelFilter, level)
-		}
+	return createWindowsEventFilter(path, levels)
+}
+
+func CreateRangeQuery(path string, levels []string, r state.Range) (*uint16, error) {
+	rangeFilter := fmt.Sprintf(eventRangeFilter, r.StartOffset(), r.EndOffset())
+	return createWindowsEventFilter(path, levels, rangeFilter)
+}
+
+func createWindowsEventFilter(path string, levels []string, additionalFilters ...string) (*uint16, error) {
+	// Add log levels
+	var levelsFilter string
+	formattedLevels := make([]string, len(levels))
+	for i, level := range levels {
+		formattedLevels[i] = fmt.Sprintf(eventLogLevelFilter, level)
+	}
+	if len(formattedLevels) > 0 {
+		levelsFilter = fmt.Sprintf("(%s)", strings.Join(formattedLevels, " or "))
 	}
 
-	//Ignore events older than 2 weeks
+	// Ignore events older than 2 weeks
 	cutOffPeriod := (time.Hour * 24 * 14).Nanoseconds()
 	ignoreOlderThanTwoWeeksFilter := fmt.Sprintf(eventIgnoreOldFilter, cutOffPeriod/int64(time.Millisecond))
-	if filterLevels != "" {
-		filterLevels = "*[System[(" + filterLevels + ") and " + ignoreOlderThanTwoWeeksFilter + "]]"
-	} else {
-		filterLevels = "*[System[" + ignoreOlderThanTwoWeeksFilter + "]]"
+
+	var filters []string
+	if levelsFilter != "" {
+		filters = append(filters, levelsFilter)
+	}
+	filters = append(filters, ignoreOlderThanTwoWeeksFilter)
+	// Add any additional filters (e.g. record IDs)
+	if len(additionalFilters) > 0 {
+		filters = append(filters, strings.Join(additionalFilters, " and "))
 	}
 
-	xml := fmt.Sprintf(eventLogQueryTemplate, path, filterLevels)
+	xml := fmt.Sprintf(eventLogQueryTemplate, path, strings.Join(filters, " and "))
```

Comment:
Is there a limit to the query length? We're not going to hit it with the range filters, but can see it being a problem later on.